### DISCUSSION TITLE
Why implementing Handler for `FnOnce` and not only for `Fn`?

### DISCUSSION THREAD

**Discussion by Tudyx** [tokio-rs/axum#2003] [General]
Hello, I was wondering why the [handler] trait, which is a subtrait of `Clone` is implemented for `FnOnce`.

To my understanding implementing  `FnOnce` give users more flexibility, as it can also accept `FnMut` and `Fn`.
So we can use it for closure that can mutably capture their environment. But as it also requires `Clone`, we can't mutably capture the environment:

```rust
// If you remove the `Clone` bound then it works.
fn execute_closure<F: FnOnce() + Clone>(func: F) {
    func()
}

fn main() {
    let mut count = 0;
    execute_closure(|| {
        println!("hello world!");
        count += 1;
    });
}
```
[playground]

So what is the point to implement  `FnOnce` if we can't really mutably capture the environment?  Why not implement only for `Fn`?

So they might be another reason, but I can't figure it out, if you could please enlighten me.

[handler]: https://docs.rs/axum/latest/axum/handler/trait.Handler.html
[playground]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=972af6bfb871c818049546ca210358f7

**Comment by traitman** [comment]
see https://doc.rust-lang.org/std/ops/trait.FnOnce.html

Since both [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html) and [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) are subtraits of FnOnce, any instance of [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html) or [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) can be used where a FnOnce is expected.

so, if you expect the param to be `FnOnce`, user can also pass `Fn`,  `FnMut` to it.

**Comment by Tudyx** [comment]
Thanks for your reply!

> Since both [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html) and [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) are subtraits of [FnOnce](https://doc.rust-lang.org/std/ops/trait.Fn.html), any instance of Fn or [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) can be used where a FnOnce is expected.

That's what I mentioned at the beginning of my question. My point is why require `FnOnce` **and** `Clone`, because you can't really take benefit of `FnMut`  in that case (like mutably capture the environment, see [playground]).


 [playground]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=972af6bfb871c818049546ca210358f7

**Comment by davidpdrsn** [comment]
I don't specifically remember but would a fun exercise to change it to `Fn` (`FnMut`?) and see what happens.

**Comment by ttys3** [comment]
maybe related: https://github.com/hyperium/hyper/pull/3223

hyper just changed service `&mut self` to  `&self`

`F: FnMut(Request<ReqBody>) -> Ret` =>  `F: Fn(Request<ReqBody>) -> Ret`

https://github.com/hyperium/hyper/commit/d894439e009aa75103f6382a7ba98fb17da72f02
